### PR TITLE
[SPARK-9241] [SQL] [WIP] Supporting multiple DISTINCT columns

### PR DESCRIPTION
**Pull request by hvanhovell**
This PR adds support for multiple distinct columns to the new aggregation code path. 

The implementation uses the `OpenHashSet` class and set expressions. As a result we can only use the slower sort based aggregation code path. This also means the code will be probably slower than the old hash aggregation.

The PR is currently in the proof of concept phase, and I have submitted it to get some feedback to see if I am headed in the right direction. I'll add more tests if this considered to be the way to go.

An example using the new code path:

```
val df = sqlContext
  .range(1 << 25)
  .select(
    $"id".as("employee_id"),
    (rand(6321782L) * 4 + 1).cast("int").as("department_id"),
    when(rand(981293L) >= 0.5, "M").otherwise("F").as("gender"),
    (rand(7123L) * 3 + 1).cast("int").as("education_level")
  )

df.registerTempTable("employee")

// Regular query.
sql("""
select   department_id as d,
         count(distinct gender, education_level) as c0,
         count(distinct gender) as c1,
         count(distinct education_level) as c2
from     employee
group by department_id
""").show()
```

cc @yhuai

### CODE REVIEW

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/distinctFallback.scala [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        }
+      }
+      evaluate.eval(buffer)
+    } else null
+  }
+}
+
+case class DropAnyNull(child: Expression) extends UnaryExpression {
+  override def nullable: Boolean = false
+  override def dataType: DataType = child.dataType
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/distinctFallback.scala]: I'll remove this in the next iteration...

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/distinctFallback.scala]: Removed....

**Comment by rxin** [comment]
@hvanhovell I think the approach we want to take is to either use the aggregate expansion (#9406), or use joins, and not this one. Do you mind closing this one? (We already have a record of it from JIRA in case we need to reference it).

**Comment by hvanhovell** [comment]
Closing PR. Some of the code in here will probably re-emerge if we ever want distincts in window functions.

**Pull request closed without merging**